Allow eframe apps to be detached

rockisch opened the pull request:
* Closes <https://github.com/emilk/egui/issues/2875>.

One choice that I'm not really sure which I prefer is whether we should check for `window.id() == window_id` on `WindowEvent` events inside the `on_event` method or not. Since we're already exposing the `window`, the user could easily check that outside. However, when writing some example code, it ended up looking really bad and confusing. At the same time, people who will use this are probably people who have a good understanding of `winit` and GUI development in general, so maybe that's not an issue.

This gist shows how the 2 approaches would look like:
https://gist.github.com/rockisch/c672345661dd5411f5611b7ab14faea0

This initial version of the PR has the check inside, but I could remove it if wanted.

rockisch commented on crates/eframe/src/native/run.rs:
```diff
@@ -270,14 +271,59 @@ fn run_and_return(
+}
 
-    let mut next_repaint_time = Instant::now();
+/// Indicates what is the current state of the app after a [`Detached::on_event`] call.
+pub enum DetachedResult {
+    /// Indicates the next call to [`Detached::on_event`] will result in either a state or paint update.
+    /// Ideally, the event loop control should be set to [`winit::event_loop::ControlFlow::Poll`].
+    UpdateNext,
+    /// Indicates a call to [`Detached::on_event`] will only be required after the specified time.
+    /// Ideally, the event loop control should be set to [`winit::event_loop::ControlFlow::WaitUntil`].
+    UpdateAt(Instant),
+    /// Indicates the app received a close request.
+    Exit,
+}
```
Couldn't think of a good way to write these docs, feel free to suggest better ways.

rockisch replied:
Also, this mostly mimics `EventResult` but without `Wait` and `RepaintNow`, since that only makes sense internally. We could easily export that instead and just mention on the docs that those states should be ignored.

emilk replied:
I think this looks good!

emilk commented:
This looks very promising! Can you please add a working example too? 🙏

rockisch commented:
Added a working example and fixed the CI issues (was on a Windows machine before, so I couldn't run the `check.sh` script)

rockisch commented:
Btw, I went and changed one of my projects to this approach, and I noticed that although it does work, communication between the 'outer' code and the 'detached eframe app' is kinda clunky, since there's no way to access or modify the app data from the returned reference.

The way I'm doing this at the moment is by leaking a `RefCell`, more or less like this:
```rs
struct SharedState {
    flag: bool
}

struct MyApp {
    shared: &'static RefCell<SharedState>
}

let shared_state = Box::leak(Box::new(RefCell::new(SharedState { flag: true }))) as &_;
let detached_app = eframe::run_detached_native(
    ...,
    Box::new(|_cc| Box::new(MyApp { shared: shared_state }))
)
// 'shared_state' can be accessed and modified
```

How would you feel about exposing a `get_app(&self) -> MyApp` and `get_app_mut(&mut self) -> MyApp` on `Detached`? I took a look at the source code and I feel like I could add those with minimal changes.

CoryRobertson commented:
not sure I follow entirely but would this pr allow you to open a completely detached eframe, meaning you would have two separate event loops?

Aaron1011 commented:
What's the status of this PR? This would be very useful for Ruffle

samyak-jain commented:
> Btw, I went and changed one of my projects to this approach, and I noticed that although it does work, communication between the 'outer' code and the 'detached eframe app' is kinda clunky, since there's no way to access or modify the app data from the returned reference.
> 
> The way I'm doing this at the moment is by leaking a `RefCell`, more or less like this:
> 
> ```rust
> struct SharedState {
>     flag: bool
> }
> 
> struct MyApp {
>     shared: &'static RefCell<SharedState>
> }
> 
> let shared_state = Box::leak(Box::new(RefCell::new(SharedState { flag: true }))) as &_;
> let detached_app = eframe::run_detached_native(
>     ...,
>     Box::new(|_cc| Box::new(MyApp { shared: shared_state }))
> )
> // 'shared_state' can be accessed and modified
> ```
> 
> How would you feel about exposing a `get_app(&self) -> MyApp` and `get_app_mut(&mut self) -> MyApp` on `Detached`? I took a look at the source code and I feel like I could add those with minimal changes.

Is the problem that we need to give away ownership of the `MyApp` instance? Shouldn't the shared state be accessible when you are implementing the `update` method on `MyApp`?

Btw, if there's anyway I can contribute, I would be happy to help. 
Thanks for taking your time to implement this!

emilk commented on examples/detached_app/README.md:
```diff
@@ -0,0 +1,7 @@
+Example showing some UI controls like `Label`, `TextEdit`, `Slider`, `Button`.
```
This is not a good description

emilk replied:
And the screenshot is not very useful either

emilk commented on examples/detached_app/src/main.rs:
```diff
@@ -0,0 +1,91 @@
+    );
+    // Winit window managed by the application
+    let winit_window = winit::window::WindowBuilder::new()
+        .build(&event_loop)
```
Giving this a title of e.g. "winit window" would help explain to the user what is going on

emilk commented on examples/detached_app/src/main.rs:
```diff
@@ -0,0 +1,91 @@
+impl eframe::App for MyApp {
+    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
+        egui::CentralPanel::default().show(ctx, |ui| {
+            ui.heading("My egui Application");
+            ui.horizontal(|ui| {
+                let name_label = ui.label("Your name: ");
+                ui.text_edit_singleline(&mut self.name)
+                    .labelled_by(name_label.id);
+            });
+            ui.add(egui::Slider::new(&mut self.age, 0..=120).text("age"));
+            if ui.button("Click each year").clicked() {
+                self.age += 1;
+            }
+            ui.label(format!("Hello '{}', age {}", self.name, self.age));
```
Instead of showing the generic egui application, it would be better to show some text about what is going on in this example

emilk commented on crates/eframe/src/native/run.rs:
```diff
@@ -270,14 +271,65 @@ fn run_and_return(
+/// An eframe app detached from the event loop.
+///
+/// This is useful to run `eframe` apps while still having control over the event loop.
+/// For example, when you need to have a window managed by `egui` and another managed
```
```suggestion
/// For example, when you need to have a window managed by `eframe` and another managed
```

emilk commented on crates/eframe/src/lib.rs:
```diff
@@ -295,6 +299,94 @@ pub fn run_simple_native(
+/// Execute an app in a detached state.
+///
+/// This allows you to control the event loop itself, which is necessary in a few
+/// occasions, like when you need a screen not managed by `egui`.
```
```suggestion
/// occasions, like when you need a screen not managed by `eframe`.
```

emilk commented on crates/eframe/src/native/run.rs:
```diff
@@ -308,53 +360,97 @@ fn run_and_exit(event_loop: EventLoop<UserEvent>, mut winit_app: impl WinitApp +
+                window_id,
+                event: _event,
+            } => {
+                // Ignore window events for other windows
+                if let Some(window) = self.winit_app.window() {
```
I think you need to merge in `master` and handle the case of multi-viewports

emilk reviewed (changes requested):
Needs a merging in of `master` and testing with multi-viewports, otherwise its looking good

JackCrumpLeys commented:
This seems dead. I would be interested in reviving and finishing off the PR. Would that be OK? I am not sure how much has ganged here do you think it might be better to restart?

wpbrown commented:
Yes this is quite divergent at this point. I have a new approach off master here that could use some review. #6750
